File: backend/app/schemas/preview_access.py
from datetime import datetime
from pydantic import BaseModel, field_validator


class PublicGateLoginRequest(BaseModel):
    email_or_username: str
    password: str


class PreviewAccessUserPublic(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    email: str
    first_name: str | None
    last_name: str | None
    is_active: bool
    note: str | None
    last_used_at: datetime | None
    created_at: datetime
    updated_at: datetime


class PreviewAccessUserCreate(BaseModel):
    email: str
    password: str
    first_name: str | None = None
    last_name: str | None = None
    is_active: bool = True
    note: str | None = None

    @field_validator('email')
    @classmethod
    def normalize_email(cls, v: str) -> str:
        return v.strip().lower()

    @field_validator('password')
    @classmethod
    def validate_password(cls, v: str) -> str:
        if len(v) < 10:
            raise ValueError('Passwort muss mindestens 10 Zeichen haben.')
        return v


class PreviewAccessUserUpdate(BaseModel):
    first_name: str | None = None
    last_name: str | None = None
    note: str | None = None


class PreviewAccessPasswordReset(BaseModel):
    new_password: str
    confirm_password: str

    @field_validator('new_password')
    @classmethod
    def validate_password(cls, v: str) -> str:
        if len(v) < 10:
            raise ValueError('Passwort muss mindestens 10 Zeichen haben.')
        return v

File: frontend/src/api/previewAccess.ts
export interface PreviewAccessUser {
  id: number
  email: string
  first_name: string | null
  last_name: string | null
  is_active: boolean
  note: string | null
  last_used_at: string | null
  created_at: string
  updated_at: string
}

export interface PreviewAccessUserCreate {
  email: string
  password: string
  first_name?: string | null
  last_name?: string | null
  is_active?: boolean
  note?: string | null
}

export interface PreviewAccessUserUpdate {
  first_name?: string | null
  last_name?: string | null
  note?: string | null
}

export interface PreviewAccessPasswordReset {
  new_password: string
  confirm_password: string
}

function authHeader(): { Authorization: string } {
  const token = localStorage.getItem('am_token')
  return { Authorization: `Bearer ${token ?? ''}` }
}

const BASE = '/api/v1/platform-admin/test-access-users'

export async function listPreviewUsers(params?: {
  search?: string
  is_active?: boolean
}): Promise<PreviewAccessUser[]> {
  const url = new URL(BASE, window.location.origin)
  if (params?.search) url.searchParams.set('search', params.search)
  if (params?.is_active !== undefined) url.searchParams.set('is_active', String(params.is_active))

  const res = await fetch(url.toString(), { headers: authHeader() })
  if (!res.ok) throw Object.assign(new Error('list_preview_users'), { response: res, status: res.status })
  return res.json()
}

export async function createPreviewUser(payload: PreviewAccessUserCreate): Promise<PreviewAccessUser> {
  const res = await fetch(BASE, {
    method: 'POST',
    headers: { ...authHeader(), 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const body = await res.json().catch(() => ({}))
    throw Object.assign(new Error('create_preview_user'), { response: res, status: res.status, body })
  }
  return res.json()
}

export async function getPreviewUser(id: number): Promise<PreviewAccessUser> {
  const res = await fetch(`${BASE}/${id}`, { headers: authHeader() })
  if (!res.ok) throw Object.assign(new Error('get_preview_user'), { response: res, status: res.status })
  return res.json()
}

export async function updatePreviewUser(id: number, payload: PreviewAccessUserUpdate): Promise<PreviewAccessUser> {
  const res = await fetch(`${BASE}/${id}`, {
    method: 'PATCH',
    headers: { ...authHeader(), 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const body = await res.json().catch(() => ({}))
    throw Object.assign(new Error('update_preview_user'), { response: res, status: res.status, body })
  }
  return res.json()
}

export async function activatePreviewUser(id: number): Promise<PreviewAccessUser> {
  const res = await fetch(`${BASE}/${id}/activate`, {
    method: 'POST',
    headers: authHeader(),
  })
  if (!res.ok) throw Object.assign(new Error('activate_preview_user'), { response: res, status: res.status })
  return res.json()
}

export async function deactivatePreviewUser(id: number): Promise<PreviewAccessUser> {
  const res = await fetch(`${BASE}/${id}/deactivate`, {
    method: 'POST',
    headers: authHeader(),
  })
  if (!res.ok) {
    const body = await res.json().catch(() => ({}))
    throw Object.assign(new Error('deactivate_preview_user'), { response: res, status: res.status, body })
  }
  return res.json()
}

export async function resetPreviewUserPassword(id: number, payload: PreviewAccessPasswordReset): Promise<void> {
  const res = await fetch(`${BASE}/${id}/reset-password`, {
    method: 'POST',
    headers: { ...authHeader(), 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const body = await res.json().catch(() => ({}))
    throw Object.assign(new Error('reset_preview_password'), { response: res, status: res.status, body })
  }
}
